### PR TITLE
[#11478] Include a list of non-submitters in the downloaded results

### LINKED ISSUE #11478: Include a list of non-submitters in the downloaded results
When an instructor views results, TEAMMATES shows a list of non-submitters at the bottom of the page.
image

But that list is not included in the csv file, if the instructor downloaded the results.

Suggestion: include the non-submitters list (name, team, email) in the the downloaded spreadsheet as well.

### PR DESCRIPTION
**Pull request by shadowezz**
Fixes #11478 

**Outline of Solution**

Add non-submitters list to output array after observable for adding question responses completes.

### CODE REVIEW

**Comment by moziliar** [comment]
@shadowezz too much reformatting. Please refrain from making too much formatting related changes that makes reviewing a lot harder.

**Comment by shadowezz** [comment]
Yup, my apologies. My VScode prettier plugin auto reformatted a bunch of stuff on save. The main change is only on line 594-604 of instructor-session-result-page.component.ts. If need be, I can try reverting the formatting changes.

**Comment by moziliar** [comment]
When it's confirmed that the last interested contributor is no longer active on this issue, you can revert the formatting related changes and ask for review from us:)

**Comment by moziliar** [comment]
Please remove all the reformatting (and if you think some of them make more sense, do make it in a separate PR without affecting code logic) before review.

**Comment by shadowezz** [comment]
@moziliar I have removed the reformatting and this PR is ready for review :)

**Review by moziliar — COMMENTED** [review]
Interesting UX decision about the progress bar. Does it mean that the progress will be stuck at 99 when we are processing all the students without submission?

Can you please provide a short gif/video on how this change works? Specifically, when there are students without submission and you try to download the csv.
A screenshot of the output csv showing the before and after change would be very helpful too for @damithc to vet the UX.

Another note: did you set isDownlodingResults to false anywhere after removing it from finalize pipe? Failure to do so will incur bug on repeated download I feel.

**Comment by shadowezz** [comment]
Yup, thats correct. I suppose processing students without submission will not take long as its just one entry per student compared to multiple responses per question. Updating the progress percentage using both questions and students without submission as base seems a little strange too. 

Here is a video of the downloading:

https://user-images.githubusercontent.com/52622254/145698832-cf183da5-edf5-4039-913f-12d7df29a33a.mov

I noticed that the progress bar doesnt really move according to the percentage completion but it is already the case before my changes.

Before csv:
Before

After csv:
After

From what I understand, isDownloadingResults is automatically handled when the model is closed so I dint have an additional toggle. Seems to be working fine. 
code

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 1]
```diff
@@ -454,22 +454,33 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
-          const progressPercentage: number = Math.round(100 * numberOfQuestionsDownloaded / totalNumberOfQuestions);
+          const progressPercentage: number = Math.round(99 * numberOfQuestionsDownloaded / totalNumberOfQuestions);
           this.progressBarService.updateProgress(progressPercentage);
         },
         complete: () => {
           if (downloadAborted) {
             return;
           }
+
+          if (this.noResponseStudents.length !== 0) {
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:474]: Try to reduce arrow anti-pattern by making this `if (this.noResponseStudents.length == 0) { return; }` like line 470-472.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 2]
```diff
@@ -454,22 +454,33 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
+            // add non-submitters list
+            out.push('Participants who have not responded to any question\n\n');
+            out.push('Team,Name\n');
+
+            for (const student of this.noResponseStudents) {
+              out.push(`${student.teamName},${student.name}\n`);
+            }
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:481]: Can consider using array foreach which means more appropriately here.

**Review by moziliar — CHANGES_REQUESTED** [review]
Looks good in terms of code. 

nit: From the recording, the progress bar seems to speed through 50-100%. Any concern here?

@damithc can I check if the UX is ok here?

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts (hunk 1 above)
Comment (shadowezz) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:474]: off the top of my head I cant really think of a good way to do this cos I still need to save the csv, close the model, etc so cant just return immediately.

Comment (moziliar) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:474]: If your if block is not executed, won't your complete callback return anyways?

Comment (moziliar) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:474]: I see what you mean here. It should stay as is then.

**Comment by shadowezz** [comment]
Screenshot 2021-12-12 at 5 14 22 PM

I noticed that you can download results from here too and the code for that seems to be largely duplicated. I can copy my changes over but I think a better way is to write some sort of a service that can be used in both pages

**Comment by moziliar** [comment]
@shadowezz that is absolutely encouraged:) you can try to make a clean update and request for review again. Any good level of abstraction is welcome!

**Comment by shadowezz** [comment]
@moziliar I have made quite a bit of changes, pls take a look. I placed all the downloading logic inside feedback-session-actions.service to reduce code duplication and make it easier to change in the future.

File: src/web/app/pages-instructor/instructor-session-base-page.component.ts [hunk 3]
```diff
@@ -229,63 +229,24 @@ export abstract class InstructorSessionBasePageComponent {
...
-          loadingModal.close();
-          blob = new Blob(outputData, { type: 'text/csv' });
-          saveAs(blob, filename);
+          this.isResultActionLoading = false;
         },
-        error: (resp: ErrorMessageOutput) => {
-          this.statusMessageService.showErrorToast(resp.error.message);
-          loadingModal.close();
+        error: () => {
+          this.isResultActionLoading = false;
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-base-page.component.ts:249]: Use piped finalize?

File: src/web/services/feedback-session-actions.service.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          // Include non-submitter list
+          const totalNumberOfItems: number = questions.length + 1;
+          const progressPercentage: number = Math.round(100 * numberOfItemsDownloaded / totalNumberOfItems);
+          this.progressBarService.updateProgress(progressPercentage);
+        },
+        complete: () => {
+          if (downloadAborted) {
+            return;
+          }
+          loadingModal.close();
```
Comment (moziliar) [inline src/web/services/feedback-session-actions.service.ts:92]: Can we use finalize for modal closing?

File: src/web/services/feedback-session-actions.service.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+          blob = new Blob(outputData, { type: 'text/csv' });
+          saveAs(blob, filename);
+        },
+        error: (resp: ErrorMessageOutput) => {
+          this.statusMessageService.showErrorToast(resp.error.message);
+          loadingModal.close();
```
Comment (moziliar) [inline src/web/services/feedback-session-actions.service.ts:98]: ditto

**Review by moziliar — APPROVED** [review]
Sorry for the late reply. 
LGTM with some nitpicking.

@t-cheepeng you wanna take a look at the frontend changes?

**Review by t-cheepeng — APPROVED** [review]
Frontend changes looks good.

@damithc Please help to check if the progress bar is good

**Comment by damithc** [comment]
> Frontend changes looks good.
> 
> @damithc Please help to check if the progress bar is good

sorry, I missed this before. I thought the progress bar existed even before? I'm fine as long as it shows show progress while the results are downloading.

@shadowezz can you also include the email address column of the students in the csv file? That could be useful if the instructor wishes to send an email to those students.

**Comment by shadowezz** [comment]
>  I thought the progress bar existed even before? I'm fine as long as it shows show progress while the results are downloading.

Yea the progress bar has always been there. I just included the non-submitter section when calculating progress.

> can you also include the email address column of the students in the csv file? That could be useful if the instructor wishes to send an email to those students.

Sure

**Comment by shadowezz** [comment]
Screenshot 2022-01-14 at 10 58 05 AM

@damithc Looks like this now

**Comment by damithc** [comment]
> Screenshot 2022-01-14 at 10 58 05 AM
> 
> @damithc Looks like this now

Looks good @shadowezz

**Review by damithc — APPROVED** [review]
I'm OK with the external behavior. Did not review the code though.

File: src/e2e/java/teammates/e2e/cases/InstructorFeedbackReportPageE2ETest.java [hunk 6]
```diff
@@ -331,7 +331,7 @@ public void testActions() {
         resultsPage = loginToPage(resultsUrl, InstructorFeedbackResultsPage.class, instructor.getGoogleId());
 
         ______TS("verify loaded session details");
-        resultsPage.verifySessionDetails(feedbackSession);
+        // resultsPage.verifySessionDetails(feedbackSession);
```
Comment (damithc) [inline src/e2e/java/teammates/e2e/cases/InstructorFeedbackReportPageE2ETest.java:334]: The code needs further cleaning :-)

Comment (shadowezz) [reply src/e2e/java/teammates/e2e/cases/InstructorFeedbackReportPageE2ETest.java]: oh dear my bad

**Comment by moziliar** [comment]
Code LGTM. Good job!

**Pull request merged**